File: src/main/java/no/fintlabs/mapping/selectable/EventCategorySelectableMappingService.java
package no.fintlabs.mapping.selectable;

import no.fintlabs.model.Selectable;
import no.fintlabs.model.event.EventCategory;
import no.fintlabs.model.instance.InstanceStatus;
import org.springframework.stereotype.Service;

@Service
public class EventCategorySelectableMappingService {

    private final InstanceStatusSelectableMappingService instanceStatusSelectableMappingService;

    public EventCategorySelectableMappingService(
            InstanceStatusSelectableMappingService instanceStatusSelectableMappingService
    ) {
        this.instanceStatusSelectableMappingService = instanceStatusSelectableMappingService;
    }

    public Selectable<String> toSelectable(EventCategory eventCategory) {
        return Selectable
                .<String>builder()
                .value(eventCategory.name())
                .label(getDisplayText(eventCategory))
                .build();
    }

    private String getDisplayText(EventCategory eventCategory) {
        return switch (eventCategory) {
            case INSTANCE_RECEIVED -> "Mottatt";
            case INSTANCE_REGISTERED -> "Mellomlagret";
            case INSTANCE_REQUESTED_FOR_RETRY -> "Forespurt for nytt forsøk";
            case INSTANCE_MAPPED -> "Konvertert";
            case INSTANCE_READY_FOR_DISPATCH -> "Klar for sending til destinasjon";
            case INSTANCE_DISPATCHED -> "Sendt til destinasjon";
            case INSTANCE_MANUALLY_PROCESSED -> "Manuelt behandlet";
            case INSTANCE_MANUALLY_REJECTED -> "Manuelt avvist";
            case INSTANCE_STATUS_OVERRIDDEN_AS_TRANSFERRED ->
                    "Status manuelt overstyrt som '" + instanceStatusSelectableMappingService.getDisplayText(InstanceStatus.TRANSFERRED) + "'";
            case INSTANCE_RECEIVAL_ERROR -> "Feilet under mottak";
            case INSTANCE_REGISTRATION_ERROR -> "Feilet under registrering";
            case INSTANCE_RETRY_REQUEST_ERROR -> "Feilet under forespørsel om nytt forsøk";
            case INSTANCE_MAPPING_ERROR -> "Feilet under konvertering";
            case INSTANCE_DISPATCHING_ERROR -> "Feilet under sending til destinasjon";
            case INSTANCE_DELETED -> "Mellomlagring av instans slettet";
        };
    }

}

File: src/main/kotlin/no/novari/flyt/history/mapping/selectable/EventCategorySelectableMappingService.kt
package no.novari.flyt.history.mapping.selectable

import no.novari.flyt.history.model.Selectable
import no.novari.flyt.history.model.event.EventCategory
import no.novari.flyt.history.model.instance.InstanceStatus
import org.springframework.stereotype.Service

@Service
class EventCategorySelectableMappingService(
    private val instanceStatusSelectableMappingService: InstanceStatusSelectableMappingService,
) {
    fun toSelectable(eventCategory: EventCategory): Selectable<String> {
        return Selectable(
            value = eventCategory.name,
            label = getDisplayText(eventCategory),
        )
    }

    private fun getDisplayText(eventCategory: EventCategory): String {
        return when (eventCategory) {
            EventCategory.INSTANCE_RECEIVED -> {
                "Mottatt"
            }

            EventCategory.INSTANCE_REGISTERED -> {
                "Mellomlagret"
            }

            EventCategory.INSTANCE_REQUESTED_FOR_RETRY -> {
                "Forespurt for nytt forsøk"
            }

            EventCategory.INSTANCE_MAPPED -> {
                "Konvertert"
            }

            EventCategory.INSTANCE_READY_FOR_DISPATCH -> {
                "Klar for sending til destinasjon"
            }

            EventCategory.INSTANCE_DISPATCHED -> {
                "Sendt til destinasjon"
            }

            EventCategory.INSTANCE_MANUALLY_PROCESSED -> {
                "Manuelt behandlet"
            }

            EventCategory.INSTANCE_MANUALLY_REJECTED -> {
                "Manuelt avvist"
            }

            EventCategory.INSTANCE_STATUS_OVERRIDDEN_AS_TRANSFERRED -> {
                "Status manuelt overstyrt som '${instanceStatusSelectableMappingService.getDisplayText(
                    InstanceStatus.TRANSFERRED,
                )}'"
            }

            EventCategory.INSTANCE_RECEIVAL_ERROR -> {
                "Feilet under mottak"
            }

            EventCategory.INSTANCE_REGISTRATION_ERROR -> {
                "Feilet under registrering"
            }

            EventCategory.INSTANCE_RETRY_REQUEST_ERROR -> {
                "Feilet under forespørsel om nytt forsøk"
            }

            EventCategory.INSTANCE_MAPPING_ERROR -> {
                "Feilet under konvertering"
            }

            EventCategory.INSTANCE_DISPATCHING_ERROR -> {
                "Feilet under sending til destinasjon"
            }

            EventCategory.INSTANCE_DELETED -> {
                "Mellomlagring av instans slettet"
            }
        }
    }
}
